imsim/sky_model.py

```diff
@@ -108,8 +112,130 @@ def __call__(self, x, y):
+        """
+        Parameters
+        ----------
+        img_wcs : 
+            WCS of the current image.
+        boresight : `array`
+            Array stores the center RA/DEC of boresight.
+        seed : `int`
+            Random seed number for each CCD. 
+            This number should be the det_num for the current sensor
+        spatial_vary : `bool`
+            Parameter controls whether to account for Sky line variation
+            in the fringing model or not. Default is set to True.
+        """
+        self.img_wcs = img_wcs
+        self.boresight = boresight
+        self.spatial_vary = spatial_vary
+        self.seed = seed
+        
+    def generate_heightfield(self,fractal_dimension=2.5, n=4096):
+        '''
+        # Use the spectral sythesis method to generate a heightfield
+        '''
+        H = 1 - (fractal_dimension - 2)
+        kpow = -(H + 1.0) / 1.2
+        
+        A = np.zeros((n, n), complex)
+
+        kvec = np.fft.fftfreq(n)
+        k0 = kvec[n // 64]
+        kx, ky = np.meshgrid(kvec, kvec, sparse=True, copy=False)
+        ksq = kx ** 2 + ky ** 2
+        m = ksq > 0
+        random_seed = self.seed
+        gen = galsim.BaseDeviate(random_seed).np
+        phase = 2 * np.pi * gen.uniform(size=(n, n))
+        A[m] = ksq[m] ** kpow * gen.normal(size=(n, n))[m] * np.exp(1.j * phase[m]) * np.exp(-ksq[m] / k0 ** 2)
+
+        return np.fft.ifft2(A)
 
 
+    def simulate_fringes(self, amp=0.002):
+        '''
+        # Generate random fringing pattern from a heightfield
+        Parameters
+        ------------------------
+        amp: `float`
+            Fringing amplitude. Default is set to 0.002 (0.02%)
+        '''
+        n = 1.2
+        n1 = 1.5
+        nwaves_rms=10.
+        X = self.generate_heightfield(n, 4096)
+        X *= nwaves_rms / np.std(X.real)
+        Z =  amp * np.cos(2 * n1 * (X.real))
+        # Scale up to unity
+        Z  += 1
+        return(Z)
+
+        
+    def fringe_variation_level(self):
+        '''
+        Function implementing temporal and spatial variation of fringing. 
+        
+        The function will return a multiplicative factor that modifies
+        the fringing amplitude from sensor to sensor based on the 
+        time (not implemented yet) and its location on the focal plane.
+        
+        Set spatial = True to turn on spatial variation implementation.
+        Otherwise, this function will return a unity value.
+        '''
+        
+        if self.spatial_vary:
+            # Load 2d interpolator for OH spatial variation
+            filename = os.path.join(data_dir, 'fringing_data',
+                                        'skyline_var.pkl')
+            with open(filename, 'rb') as f:
+                interp = pickle.load(f)
+                
+            dx = self.boresight[0] - self.img_wcs.center.ra.deg 
+            dy = self.boresight[1] - self.img_wcs.center.dec.deg
```

Comment:
If boresight is a CelestialCoord, then this becomes:
```
dx, dy = self.boresight.project(self.img_wcs.toWorld(image.true_center))
dx = dx.deg
dy = dy.deg
```
(Although I would delay getting the degree versions until you use them in the line below.)

This has the added advantage that it would work with any wcs, rather than specifically the batoid wcs.  (The `center` attribute is an implementation detail of that wcs type.)

Although note that this function currently doesn't have access to the `image`, so you probably want to compute and store the combination `image.wcs.toWorld(image.true_center)` in the constructor rather than store `img_wcs` as you currently do.

Reply:
Oh, and it also has the other advantage that the scale in the RA direction doesn't change as you move far from the equator.  I'm pretty sure that must be a bug in the current implementation.

Reply:
That's a bug in BatoidWCS?  What's the bug?

Reply:
Not in BatoidWCS.  It's a bug in these lines.  If you're near the north pole, delta RA is much larger than the actual angle on the sky.  There's a cos(dec) factor that is neglected here, but which `project` does correctly.

Reply:
Actually, thinking about this more, does this imply that you have the same bug in the way you made the interpolation file?  If you based this on observations of fringing data, then I think this means you would have hard-coded a cos(dec) factor in there corresponding to whatever declination the observations were made.

Reply:
I talked to Chris about my above concern, and I think the interpolation is probably fine.  But you should switch to using project for the position (dx,dy) you use for calling the interpolation.